README.md

```diff
@@ -1 +1,247 @@
-# kernel-module-imx-pwm-led
+# Freescale i.MX PWM LED Linux Kernel Module
+This Linux loadable kernel module controls multiple PWM channels for driving
+LEDs using the SDMA on Freescale (NXP) [i.MX](https://www.nxp.com/products/processors-and-microcontrollers/arm-processors/i-mx-applications-processors)
+processors.
+
+The module supports pre-loading of 'fades' (a buffer of PWM duty-cycle samples)
+and 'cues' (a list of actions to perform on one or more PWM channels), that can
+then be quickly triggered later using `ioctl`.
+
+Much of this code is adapted from the [imx-pwm-audio](https://github.com/Glowforge/kernel-module-imx-pwm-audio/)
+module from Glowforge and the [meta-openglow](https://github.com/ScottW514/meta-openglow)
+kernel patches from OpenGlow. The [i.MX SDMA assembler](http://billauer.co.il/blog/2011/10/imx-sdma-howto-assembler-linux/)
+from Eli Billauer, and [hints](https://community.nxp.com/thread/356855) from
+TomE for working round a hardware issue in the i.MX were also used.
+
+## Quick-Start Guide
+Using [ioctl](https://github.com/jerome-pouiller/ioctl/):
+```
+# Load the module with parameters:
+insmod imx_pwm_led.ko pwm_period=48000 pwm_invert=0
+
+# Open fade0:
+ioctl /dev/pwm_led0 0x00007541 -v 0
+
+# Write a fade file:
+cat my_fade.bin > /dev/pwm_led0
+
+# Open cue0:
+ioctl /dev/pwm_led0 0x00007542 -v 0
+
+# Write 2 cue actions to play fade0 on channels 0 and 1:
+printf "\x00\x00\x00\x00\x01\x00\x00\x00" > /dev/pwm_led0
+
+# Play cue0:
+ioctl /dev/pwm_led0 0x00007546 -v 0
+```
+
+## Requirements
+The module requires kernel patches to expose driver APIs to the i.MX EPIT and
+SDMA. See the `linux-imx-x.x.x` directory.
+
+## User-Space API
+The module creates devices for each PWM channel visible at `/dev/pwm_led0` to
+`/dev/pwm_led7` for the iMX6 Ultralight.
+
+### Definitions
+- A PWM 'period' is defined as a 16-bit PWM clock divider. For example, if the
+  IPG clock is 24 MHz and the PWM period is 12000, the output PWM frequency
+  will be 2 kHz.
+- A PWM 'duty-cycle value' is defined as a 16-bit counter value at which the
+  output changes from 1 to 0, which can range from zero to the PWM period value.
```

Comment:
> the output changes from 1 to 0
Does that mean the outputs are normally on by default?

Reply:
I mean the output changes from 0 to 1 when the counter wraps round from pwm_period - 1 to zero, and 1 to 0 when it passes the duty cycle value. Do you think it's not clear enough? I can add this to make it clearer.

Reply:
Got it now. I guess it's clear enough.